### PR TITLE
feat(token-providers): add token based authentication

### PR DESCRIPTION
**Pull request by trivikr**
### Issue
Internal JS-3540

### Description
Adds token based authentication

### Testing
Internal JS-3173

---
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/middleware-token/README.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# @aws-sdk/middleware-token
+
+[![NPM version](https://img.shields.io/npm/v/@aws-sdk/middleware-token/latest.svg)](https://www.npmjs.com/package/@aws-sdk/middleware-token)
+[![NPM downloads](https://img.shields.io/npm/dm/@aws-sdk/middleware-token.svg)](https://www.npmjs.com/package/@aws-sdk/middleware-token)
+
+Middleware and Plugin for setting token authentication.
```
Comment (AllanZhengYP) [inline packages/middleware-token/README.md:6]: Internal package note?

Comment (AllanZhengYP) [reply packages/middleware-token/README.md:6]:
```suggestion
Middleware and Plugin for setting token authentication.

> An internal package

## Usage

You probably shouldn't, at least directly.
```

File: packages/middleware-token/src/tokenMiddleware.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  toMiddleware: "retryMiddleware",
+  override: true,
+};
+
+export const tokenMiddleware =
+  <Input extends object, Output extends object>(
+    options: TokenResolvedConfig
+  ): FinalizeRequestMiddleware<Input, Output> =>
+  (next: FinalizeHandler<Input, Output>): FinalizeHandler<Input, Output> =>
+    async function (args: FinalizeHandlerArguments<Input>): Promise<FinalizeHandlerOutput<Output>> {
```
Comment (AllanZhengYP) [inline packages/middleware-token/src/tokenMiddleware.ts:25]: super nit: arrow function?
```suggestion
    (args: FinalizeHandlerArguments<Input>): Promise<FinalizeHandlerOutput<Output>> => {
```

**Review by AllanZhengYP — COMMENTED** [review]
Looks good, only have very minor feedbacks

File: packages/middleware-token/src/tokenMiddleware.ts (hunk 2 above)
Comment (trivikr) [reply packages/middleware-token/src/tokenMiddleware.ts]: Done in [`b428374` (#3883)](https://github.com/aws/aws-sdk-js-v3/pull/3883/commits/b428374528a06ef7e4bc42eed9c80fe3b92a133e)

**Review by AllanZhengYP — APPROVED** [review]

**Pull request merged**